### ISSUE TITLE
Cannot connect to vm service on real device when the Flutter update to 2.x

### ISSUE THREAD

**Issue by talisk** [flutter/flutter#84069]
## Steps to Reproduce

**IMPORTANT** Make sure your flutter version is 2.x, I reproduced the problem on 2.0.1 stable and 2.2.1 stable. But everything works fine on the old version like 1.22.3 stable. AND everything works fine on the simulator.

[Demo](https://github.com/talisk/vmservice_demo)

1. Clone my demo project
2. `flutter pub get`
3. `flutter run`: Run the demo on iPhone real device, Android real device and simulator.
4. Press the button in the demo, check debug output

**Expected results:** The connection on Flutter 2.x is normal like the Flutter 1.22.3.

**Actual results:**

|    | iOS device  | Android device | Simulator  |
|  ----  | ----  | ----  | ----  |
| Flutter 1.22.3  | ✅ | ✅ | ✅ |
| Flutter 2.0.1  | ❌ | ❌ | ✅ |
| Flutter 2.2.1  | ❌ | ❌ | ✅ |

## Details

iOS real device output (iPhone 7 Plus with iOS 12.4, iPhone 11 Pro with iOS 14.0):
```
[VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: SocketException: OS Error: Connection refused, errno = 61, address = 127.0.0.1, port = 63737
```

Android real device output (Samsung S10 with Android 11):
```
E/flutter (21242): [ERROR:flutter/lib/ui/ui_dart_state.cc(199)] Unhandled Exception: SocketException: OS Error: Connection refused, errno = 111, address = 127.0.0.1, port = 57704
```

This is the core code:
``` dart
ServiceProtocolInfo info = await Service.getInfo();
String url = info.serverUri.toString();
Uri uri = Uri.parse(url);
Uri socketUri = convertToWebSocketUrl(serviceProtocolUrl: uri);
final isolateId = Service.getIsolateID(Isolate.current);
final service = await vmServiceConnectUri(socketUri.toString());
```

**Comment by talisk** [comment]
This is the core code:

``` dart
ServiceProtocolInfo info = await Service.getInfo();
String url = info.serverUri.toString();
Uri uri = Uri.parse(url);
Uri socketUri = convertToWebSocketUrl(serviceProtocolUrl: uri);
final isolateId = Service.getIsolateID(Isolate.current);
final service = await vmServiceConnectUri(socketUri.toString());
```

**Comment by TahaTesser** [comment]
Hi @talisk 
If you create a new project, does it still reproduce the issue? See https://github.com/flutter/flutter/issues/83978

**Comment by talisk** [comment]
> Hi @talisk
> If you create a new project, does it still reproduce the issue? See #83978

Yes, it will reproduce. The [Demo](https://github.com/talisk/vmservice_demo) which was I newly created by Flutter 2.2.1.

**Comment by TahaTesser** [comment]
Hi @talisk 
Thanks for the update


I can confirm it breaks on the latest `stable`

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.2.1, on macOS 11.4 20F71 darwin-x64, locale
    en-GB)
    • Flutter version 2.2.1 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 02c026b03c (12 days ago), 2021-05-27 12:24:44 -0700
    • Engine revision 0fdb562ac8
    • Dart version 2.13.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode-beta.app/Contents/Developer
    • Xcode 13.0, Build version 13A5154h
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.56.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[✓] Connected device (3 available)
    • IN2011 (mobile) • c9d8ee0c • android-arm64  • Android 11 (API 30)
    • macOS (desktop) • macos    • darwin-x64     • macOS 11.4 20F71 darwin-x64
    • Chrome (web)    • chrome   • web-javascript • Google Chrome 91.0.4472.77

• No issues found!
```

</details>

but `SocketException` isn't thrown on `1.22.3`

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 1.22.3, on macOS 11.4 20F71, locale en-GB)
    • Flutter version 1.22.3 at /Users/tahatesser/Code/flutter
    • Framework revision 8874f21e79 (7 months ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.0)
    • Xcode at /Volumes/Extreme/Xcode-beta.app/Contents/Developer
    • Xcode 13.0, Build version 13A5154h
    • CocoaPods version 1.10.1

[!] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.56.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[✓] Connected device (1 available)
    • IN2011 (mobile) • c9d8ee0c • android-arm64 • Android 11 (API 30)

! Doctor found issues in 1 category.
```

</details>

**Comment by jonahwilliams** [comment]
I believe this is related to https://github.com/dart-lang/sdk/issues/46256

FYI @bkonyi

**Comment by talisk** [comment]
> is there any particular reason why you're trying to connect to the VM service from the application itself?

I developed a tool for in-app debugging, similar to DevTools but only run in-app, and I will contribute it to the community soon.

This tool works well until it becomes unavailable in Flutter 2.x because of this issue.

**Comment by talisk** [comment]
> > I believe this is related to [dart-lang/sdk#46256](https://github.com/dart-lang/sdk/issues/46256)
> > FYI @bkonyi
> 
> Almost certainly. The workaround here is to pass `--no-dds` when launching the application.
> 
> @talisk is there any particular reason why you're trying to connect to the VM service from the application itself?

Is passing `--no-dds` a normally and stable way to resolve the problem? 

Before the PR https://github.com/flutter/flutter/pull/80900 , the param named `--disable-dds`? And I tried `--disable-dds` on Flutter 2.2.1 stable, it was able to connect the vm service.

**Comment by bkonyi** [comment]
> > > I believe this is related to [dart-lang/sdk#46256](https://github.com/dart-lang/sdk/issues/46256)
> > > FYI @bkonyi
> > 
> > 
> > Almost certainly. The workaround here is to pass `--no-dds` when launching the application.
> > @talisk is there any particular reason why you're trying to connect to the VM service from the application itself?
> 
> Is passing `--no-dds` a normally and stable way to resolve the problem?
> 
> Before the PR #80900 , the param named `--disable-dds`? And I tried `--disable-dds` on Flutter 2.2.1 stable, it was able to connect the vm service.

Yes, `--no-dds` should be stable and `--disable-dds` is valid for older versions without the `--no-dds` flag.

I'm interested in your use case for accessing the VM service on device. Would it be possible to share the project you're working on so I can get a better idea of how you're using the service? We've recently had discussions about connecting to a VM service running in the same application while DDS (the Dart Development Service) is already connected as the VM service refuses direct connections when DDS is connected and it's currently not possible to connect to DDS from an application since we'd need some extra port forwarding. Getting some insights into how users are utilizing the VM service would be extremely useful for these discussions.

**Comment by talisk** [comment]
> > > > I believe this is related to [dart-lang/sdk#46256](https://github.com/dart-lang/sdk/issues/46256)
> > > > FYI @bkonyi
> > > 
> > > 
> > > Almost certainly. The workaround here is to pass `--no-dds` when launching the application.
> > > @talisk is there any particular reason why you're trying to connect to the VM service from the application itself?
> > 
> > 
> > Is passing `--no-dds` a normally and stable way to resolve the problem?
> > Before the PR #80900 , the param named `--disable-dds`? And I tried `--disable-dds` on Flutter 2.2.1 stable, it was able to connect the vm service.
> 
> Yes, `--no-dds` should be stable and `--disable-dds` is valid for older versions without the `--no-dds` flag.
> 
> I'm interested in your use case for accessing the VM service on device. Would it be possible to share the project you're working on so I can get a better idea of how you're using the service? We've recently had discussions about connecting to a VM service running in the same application while DDS (the Dart Development Service) is already connected as the VM service refuses direct connections when DDS is connected and it's currently not possible to connect to DDS from an application since we'd need some extra port forwarding. Getting some insights into how users are utilizing the VM service would be extremely useful for these discussions.

My tool is a Flutter package which build an in-app debug platform. It wraps vm_service, and exposes some interfaces with certain restrictions. Any Flutter app could use the in-app tools for debugging. User could develop their custom in-app debug kits and Integrate into the in-app debug platform. 

With the tool, user could inspect their widgets, debug memory issues, check performance panel in their host app, rather than connecting to a DevTools.

**Comment by talisk** [comment]
> > > > I believe this is related to [dart-lang/sdk#46256](https://github.com/dart-lang/sdk/issues/46256)
> > > > FYI @bkonyi
> > > 
> > > 
> > > Almost certainly. The workaround here is to pass `--no-dds` when launching the application.
> > > @talisk is there any particular reason why you're trying to connect to the VM service from the application itself?
> > 
> > 
> > Is passing `--no-dds` a normally and stable way to resolve the problem?
> > Before the PR #80900 , the param named `--disable-dds`? And I tried `--disable-dds` on Flutter 2.2.1 stable, it was able to connect the vm service.
> 
> Yes, `--no-dds` should be stable and `--disable-dds` is valid for older versions without the `--no-dds` flag.
> 
> I'm interested in your use case for accessing the VM service on device. Would it be possible to share the project you're working on so I can get a better idea of how you're using the service? We've recently had discussions about connecting to a VM service running in the same application while DDS (the Dart Development Service) is already connected as the VM service refuses direct connections when DDS is connected and it's currently not possible to connect to DDS from an application since we'd need some extra port forwarding. Getting some insights into how users are utilizing the VM service would be extremely useful for these discussions.

Why does Flutter 2.x(I tried Flutter 2.0.1, 2.2.1) will reproduce the issue, but Flutter 1.22 won't? What breaking change was introduced in Flutter 2.x

**Comment by Ibtesam-Mahmood** [comment]
@bkonyi any updates on this?

**Comment by bkonyi** [comment]
No update on this as there's higher priority work on my queue at this point. The workaround here is to pass `--no-dds` to disable the Dart Development Service which will allow for applications to connect to their own VM service, as it's the DDS instance running off of device that it's not possible to connect to.

**Comment by talisk** [comment]
> > > > I believe this is related to [dart-lang/sdk#46256](https://github.com/dart-lang/sdk/issues/46256)
> > > > FYI @bkonyi
> > > 
> > > 
> > > Almost certainly. The workaround here is to pass `--no-dds` when launching the application.
> > > @talisk is there any particular reason why you're trying to connect to the VM service from the application itself?
> > 
> > 
> > Is passing `--no-dds` a normally and stable way to resolve the problem?
> > Before the PR #80900 , the param named `--disable-dds`? And I tried `--disable-dds` on Flutter 2.2.1 stable, it was able to connect the vm service.
> 
> Yes, `--no-dds` should be stable and `--disable-dds` is valid for older versions without the `--no-dds` flag.
> 
> I'm interested in your use case for accessing the VM service on device. Would it be possible to share the project you're working on so I can get a better idea of how you're using the service? We've recently had discussions about connecting to a VM service running in the same application while DDS (the Dart Development Service) is already connected as the VM service refuses direct connections when DDS is connected and it's currently not possible to connect to DDS from an application since we'd need some extra port forwarding. Getting some insights into how users are utilizing the VM service would be extremely useful for these discussions.

Hi, I'm glad to introduce my in-app debug toolkits - [flutter_ume](https://github.com/bytedance/flutter_ume). It accesses VM service on device.
